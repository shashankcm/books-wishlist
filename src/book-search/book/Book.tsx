import React, { memo, useCallback } from "react";
import { ActionTypes, Book as BookType } from "../../shared/types";
import { useStoreProviderValue } from "../../store/store";

const Book = memo(
  ({
    volumeInfo: {
      title,
      authors,
      publisher,
      publishedDate,
      description,
      imageLinks,
    },
    id,
  }: BookType) => {
    const {
      state: { wishlist },
      dispatch,
    } = useStoreProviderValue();

    const onAddToWishList = useCallback(() => {
      dispatch({
        type: ActionTypes.ADD_TO_WISHLIST,
        payload: { id, title },
      });
    }, [dispatch, id, title]);

    const shouldDisable = () => wishlist.some((items) => items.id === id);

    return (
      <div className="book__container" key={id}>
        <div className="book__container--header">
          <img
            src={
              imageLinks?.thumbnail
                ? imageLinks?.thumbnail
                : "https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2F100scopenotes.com%2Ffiles%2F2016%2F02%2Fno_book_cover_lg.jpg&f=1&nofb=1"
            }
            alt={title}
            className="book__container--header__img"
          />
          <div className="book__container--header__info">
            <div>
              <label htmlFor="Title">Title: </label>
              <div
                className="book__container--header__info--title"
                data-testid="title"
              >
                {title}
              </div>
            </div>
            <div>
              <label>Authors: </label>
              <div
                className="book__container--header__info--author"
                data-testid="author"
              >
                {authors && authors.map((author) => `${author}, `)}
              </div>
            </div>
            <div>
              <label>Publisher: </label>
              <div
                className="book__container--header__info--publisher"
                data-testid="publisher"
              >
                {publisher}
              </div>
            </div>
            <div>
              <label>Published Date: </label>
              <div
                className="book__container--header__info--published"
                data-testid="published"
              >
                {publishedDate}
              </div>
            </div>
          </div>
        </div>
        <div className="book__container--description">{description}</div>
        <div className="book__container--button__container">
          <button
            type="button"
            onClick={onAddToWishList}
            disabled={shouldDisable()}
            data-testid="addToWishList"
          >
            Add to wishlist
          </button>
        </div>
      </div>
    );
  }
);
export default Book;
