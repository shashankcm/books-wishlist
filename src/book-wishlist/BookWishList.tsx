import React from "react";
import { useStoreProviderValue } from "../store/store";

const BookWishList = () => {
  const {
    state: { wishlist },
  } = useStoreProviderValue();

  const renderWishlistedBooks = () =>
    wishlist.map(({ id, title }) => <li key={id}>{title}</li>);

  return (
    <div className="book__wishlist">
      <h3 className="book__wishlist--reading__list">
        My Reading Wishlist (
        <span data-testid="numberOfWishlistBooks">{wishlist.length || 0}</span>)
      </h3>
      <div className="book__wishlist--selected__book" data-testid="listOfBooks">
        <ol>{renderWishlistedBooks()}</ol>
      </div>
    </div>
  );
};

export default BookWishList;
