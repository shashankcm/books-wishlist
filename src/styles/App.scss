@import "./variables";
@import "./Book.scss";

* {
  box-sizing: border-box;
}

body {
  font-family: "Architects Daughter", cursive;
}

h1,
h2 {
  font-size: 24px;
  margin: $spacing-xxs 0;
  font-weight: 300;
}

form {
  margin-bottom: $spacing-xs;
}

input:focus {
  outline: none;
}

input {
  border: none;
}

main {
  padding: $spacing-xs $spacing-xxs;
  display: flex;
  height: 100vh;

  @include md {
    flex-direction: column;
  }
}

.book__search {
  flex: 0.8;
  margin: 20px;

  &--search__form {
    width: 100%;
    border: 2px solid black;
    padding: 20px;
  }

  &--search__input {
    flex: 1;
    width: 100%;
  }

  &--container {
    display: grid;
    width: 100%;
    overflow-y: scroll;
    grid-template-columns: 1fr;
    grid-template-rows: 1fr;
    border: 2px solid black;
    padding: 20px;
    min-height: 83%;

    &::-webkit-scrollbar {
      display: none;
    }

    @include lg {
      min-height: 83%;
    }

    @include sm {
      padding: 5px;
    }
  }
}

.book__wishlist {
  flex: 0.2;
  min-width: 300px;
  margin: 0 20px;

  @include md {
    margin: 60px 20px;
  }

  @include md {
    margin: 60px 20px;
  }

  &--reading__list {
    border: 2px solid black;
    text-align: center;
    -webkit-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
    -moz-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
    box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
  }

  &--selected__book {
    width: 100%;
    height: 90%;
    border: 2px solid black;
    padding: 20px;
  }
}
