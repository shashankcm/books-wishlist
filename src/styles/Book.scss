@import "./variables.scss";

.book__container {
  flex-direction: column;
  margin: 3% 2%;
  cursor: pointer;
  box-shadow: 10px 10px 16px -6px rgba(0, 0, 0, 0.75);

  &--header {
    display: grid;
    width: 100%;
    grid-template-columns: 1fr 2fr;
    grid-template-rows: 20vh;

    @include md {
      grid-template-columns: 1fr 2fr;
      grid-template-rows: 20vh;
    }

    @include sm {
      grid-template-columns: 2fr;
      grid-gap: 40px;
    }

    &__img {
      height: 100%;
      border: 1px solid black;
      -webkit-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
      -moz-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
      box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);

      @include sm {
        margin: auto;
      }
    }

    &__info {
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      label {
        font-weight: 700;
      }

      @include sm {
        padding-left: 12px;
      }
    }
  }

  &--description {
    padding: 20px;
    font-family: fantasy;
    color: #333;

    @include sm {
      padding: 10px;
    }
  }

  &--button__container {
    text-align: right;

    button {
      display: inline-block;
      border: none;
      padding: 1rem 2rem;
      margin: 3%;
      text-decoration: none;
      background: $blue-accent;
      color: $gray05;
      font-family: sans-serif;
      font-size: 1rem;
      cursor: pointer;
      text-align: center;
      transition: background 250ms ease-in-out, transform 150ms ease;
      -webkit-appearance: none;
      -moz-appearance: none;
    }

    button:hover,
    button:focus {
      background: #0053ba;
    }

    button:focus {
      outline: 1px solid #fff;
      outline-offset: -4px;
    }

    button:active {
      transform: scale(0.99);
    }

    button:disabled {
      color: $black;
      background: $gray40;
      cursor: not-allowed;
    }
  }
}
