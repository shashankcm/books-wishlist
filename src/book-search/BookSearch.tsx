import React, { useEffect, useState, useCallback, useMemo } from "react";

import Book from "./book/Book";
import { getBooksByType } from "./book-search.service";

import { useStoreProviderValue } from "../store/store";
import { ActionTypes, AllBooks, DebounceCleanupAction } from "../shared/types";

import { randomNumber } from "../shared/utils";
import { useDebouncedFunction } from "../shared/useDebouncedFunction";

const BookSearch = () => {
  const [bookType, updateBookType] = useState("");
  const [bookTypeToSearch, updateBookTypeToSearch] = useState("");
  const { state, dispatch } = useStoreProviderValue();

  const requestBooks = useCallback(async () => {
    if (bookTypeToSearch) {
      const allBooks: AllBooks = await getBooksByType(bookTypeToSearch);
      dispatch({
        type: ActionTypes.ADD_AVAILABLE_BOOKS,
        payload: allBooks,
      });
    }
  }, [bookTypeToSearch, dispatch]);

  useEffect(() => {
    async function getAllBooks() {
      await requestBooks();
    }
    getAllBooks();
  }, [bookTypeToSearch, requestBooks]);

  const books = useMemo(() => state?.availableBooks?.items, [state]);

  const generateNumber = useMemo(() => {
    randomNumber();
  }, []);

  const debouncedSearch = useDebouncedFunction(
    updateBookTypeToSearch,
    500,
    DebounceCleanupAction.Flush
  );

  const immediateSearch = useCallback(
    (...args: Parameters<typeof updateBookType>) => {
      debouncedSearch.cancel();
      updateBookType(...args);
    },
    [debouncedSearch]
  );

  const onUpdateBookType = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = event.target;
      updateBookType(value);
      const action = value ? debouncedSearch : immediateSearch;
      action(value);
    },
    [debouncedSearch, immediateSearch]
  );

  const onSearchBooks = useCallback(
    (event) => {
      event.preventDefault();
      updateBookTypeToSearch(bookType);
      immediateSearch("");
    },
    [bookType, immediateSearch]
  );

  return (
    <div className="book__search">
      <form className="book__search--search__form" onSubmit={onSearchBooks}>
        <fieldset>
          <legend>Search Books</legend>
          <label htmlFor="book_search">Book Name</label>
          <input
            className="book__search--search__input"
            autoFocus
            name="book search"
            id="book_search"
            aria-label="bookSearch"
            type="search"
            value={bookType}
            placeholder="Search for books to add to your reading list and press Enter"
            onChange={onUpdateBookType}
          />
        </fieldset>
      </form>
      <div className="book__search--container">
        {books &&
          books.map((book, index) => (
            <Book key={index} id={generateNumber} {...book} />
          ))}
      </div>
    </div>
  );
};

export default BookSearch;
