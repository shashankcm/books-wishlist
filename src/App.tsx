import React from "react";
import "./styles/App.scss";
import BookSearch from "./book-search/BookSearch";
import BookWishList from "./book-wishlist/BookWishList";

function App() {
  return (
    <main>
      <BookSearch />
      <BookWishList />
    </main>
  );
}

export default App;
