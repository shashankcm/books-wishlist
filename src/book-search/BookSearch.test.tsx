import React from "react";
import { cleanup, fireEvent, waitForElement } from "@testing-library/react";

import BookSearch from "./BookSearch";
import { renderWithContext } from "../App.test";
import { getBooksByType } from "./book-search.service";
import Book from "./book/Book";
import BookWishList from "../book-wishlist/BookWishList";

const InputSetup = () => {
  const utils = renderWithContext(<BookSearch />);
  const input = utils.getByLabelText("bookSearch");
  return {
    input,
    ...utils,
  };
};
afterAll(cleanup);

const mockJavascriptResponseData = {
  kind: "books#volumes",
  totalItems: 1972,
  jsItems: [
    {
      kind: "books#volume",
      id: "123",
      etag: "XoIn3UCxrnA",
      selfLink: "https://www.googleapis.com/books/v1/volumes/ptiYBAAAQBAJ",
      volumeInfo: {
        title: "Javascript IS Awesome",
        authors: ["Java", "Script"],
        publisher: "Javascript",
        publishedDate: "1980-10-10",
        description: "Learn Javascript",
        imageLinks: {
          thumbnail:
            "http://books.google.com/books/content?id=ptiYBAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
        },
      },
    },
  ],
};

describe("BookSearch Component", () => {
  it("should take a snapshot", () => {
    const { asFragment } = renderWithContext(<BookSearch />);
    expect(asFragment(<BookSearch />)).toMatchSnapshot();
  });

  it("should set value to be Javascript for the input", () => {
    const { input } = InputSetup();
    fireEvent.change(input, { target: { value: "Javascript" } });
    expect(input.value).toBe("Javascript");
  });

  it("should get results when debounced search input have text 'Javascript'", async () => {
    const { input } = InputSetup();
    fireEvent.change(input, { target: { value: "Javascript" } });
    const booksData = await waitForElement(() => getBooksByType(input.value));
    expect(booksData).not.toEqual({});
  });

  it("should get results for the input word 'Javascript' when user type into the input and click enter", async () => {
    const { input } = InputSetup();
    fireEvent.change(input, { target: { value: "Javascript" } });
    fireEvent.keyDown(input, { key: "Enter", code: "Enter" });
    const booksData = await waitForElement(() => getBooksByType(input.value));
    expect(booksData).not.toEqual({});
  });
});

describe("Book Component", () => {
  it("should take a snapshot", () => {
    const { jsItems } = mockJavascriptResponseData;
    const { id, volumeInfo } = jsItems[0];
    const { asFragment } = renderWithContext(
      <Book id={+id} volumeInfo={volumeInfo} />
    );
    expect(
      asFragment(<Book id={+id} volumeInfo={volumeInfo} />)
    ).toMatchSnapshot();
  });

  it("should render Book with given data example title should be 'Javascript IS Awesome'", () => {
    const { jsItems } = mockJavascriptResponseData;
    const { id, volumeInfo } = jsItems[0];
    const { getByTestId } = renderWithContext(
      <Book id={+id} volumeInfo={volumeInfo} />
    );
    expect(getByTestId("title")).toMatchSnapshot("Javascript IS Awesome");
    expect(getByTestId("publisher")).toMatchSnapshot("Javascript");
    expect(getByTestId("published")).toMatchSnapshot("1980-10-10");
  });

  it("should add books to my reading wishlist when you click on 'Add to wishlist'", async () => {
    const { jsItems } = mockJavascriptResponseData;
    const { id, volumeInfo } = jsItems[0];
    const { getByTestId, getByRole } = renderWithContext(
      <>
        <Book id={Number(id)} volumeInfo={volumeInfo} />
        <BookWishList />
      </>
    );

    expect(getByTestId("title")).toMatchSnapshot("Javascript IS Awesome");
    fireEvent.click(getByTestId("addToWishList"));

    const booksCount = await waitForElement(() =>
      getByTestId("numberOfWishlistBooks")
    );
    expect(booksCount).toHaveTextContent("1");
  });

  it("should disabled 'Add to wishlist' button after uses added book to the wishlist", async () => {
    const { jsItems } = mockJavascriptResponseData;
    const { id, volumeInfo } = jsItems[0];
    const { getByTestId } = renderWithContext(
      <>
        <Book id={Number(id)} volumeInfo={volumeInfo} />
        <BookWishList />
      </>
    );
    const addToWishListButton = getByTestId("addToWishList");
    fireEvent.click(addToWishListButton);
    expect(addToWishListButton).toBeDisabled();
  });

  it("should not add the same book to My Reading Wishlist more than 1 times", async () => {
    const { input } = InputSetup();
    const { jsItems } = mockJavascriptResponseData;
    const { id, volumeInfo } = jsItems[0];
    const { getByTestId, getByRole } = renderWithContext(
      <>
        <Book id={Number(id)} volumeInfo={volumeInfo} />
        <BookWishList />
      </>
    );

    const addToWishListButton = getByTestId("addToWishList");
    const listOfBooks = getByTestId("listOfBooks");
    const numberOfWishlistBooks = getByTestId("numberOfWishlistBooks");

    fireEvent.change(input, { target: { value: "Javascript" } });
    fireEvent.keyDown(input, { key: "Enter", code: "Enter" });
    const jsbooksData = await waitForElement(() => getBooksByType(input.value));
    expect(jsbooksData).not.toEqual({});
    fireEvent.click(addToWishListButton);
    expect(getByTestId("title")).toMatchSnapshot("Javascript IS Awesome");
    fireEvent.click(addToWishListButton);

    const booksCount = await waitForElement(() => numberOfWishlistBooks);
    expect(booksCount).toHaveTextContent("1");
    expect(listOfBooks).toHaveTextContent("Javascript IS Awesome");
  });
});
