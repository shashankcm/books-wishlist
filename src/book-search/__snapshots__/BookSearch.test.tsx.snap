// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should take a snapshot 1`] = `
<DocumentFragment>
  <main>
    <div
      class="book__search"
    >
      <form
        class="book__search--search__form"
      >
        <fieldset>
          <legend>
            Search Books
          </legend>
          <label
            for="book_search"
          >
            Book Name
          </label>
          <input
            aria-label="bookSearch"
            class="book__search--search__input"
            id="book_search"
            name="book search"
            placeholder="Search for books to add to your reading list and press Enter"
            type="search"
            value=""
          />
        </fieldset>
      </form>
      <div
        class="book__search--container"
      />
    </div>
    <div
      class="book__wishlist"
    >
      <h3
        class="book__wishlist--reading__list"
      >
        My Reading Wishlist (
        <span
          data-testid="numberOfWishlistBooks"
        >
          0
        </span>
        )
      </h3>
      <div
        class="book__wishlist--selected__book"
        data-testid="listOfBooks"
      >
        <ol />
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Book Component should add books to my reading wishlist when you click on 'Add to wishlist': Javascript IS Awesome 1`] = `
<div
  class="book__container--header__info--title"
  data-testid="title"
>
  Javascript IS Awesome
</div>
`;

exports[`Book Component should not add the same book to My Reading Wishlist more than 1 times: Javascript IS Awesome 1`] = `
<div
  class="book__container--header__info--title"
  data-testid="title"
>
  Javascript IS Awesome
</div>
`;

exports[`Book Component should render Book with given data example title should be 'Javascript IS Awesome': 1980-10-10 1`] = `
<div
  class="book__container--header__info--published"
  data-testid="published"
>
  1980-10-10
</div>
`;

exports[`Book Component should render Book with given data example title should be 'Javascript IS Awesome': Javascript 1`] = `
<div
  class="book__container--header__info--publisher"
  data-testid="publisher"
>
  Javascript
</div>
`;

exports[`Book Component should render Book with given data example title should be 'Javascript IS Awesome': Javascript IS Awesome 1`] = `
<div
  class="book__container--header__info--title"
  data-testid="title"
>
  Javascript IS Awesome
</div>
`;

exports[`Book Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="book__container"
  >
    <div
      class="book__container--header"
    >
      <img
        alt="Javascript IS Awesome"
        class="book__container--header__img"
        src="http://books.google.com/books/content?id=ptiYBAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
      <div
        class="book__container--header__info"
      >
        <div>
          <label
            for="Title"
          >
            Title: 
          </label>
          <div
            class="book__container--header__info--title"
            data-testid="title"
          >
            Javascript IS Awesome
          </div>
        </div>
        <div>
          <label>
            Authors: 
          </label>
          <div
            class="book__container--header__info--author"
            data-testid="author"
          >
            Java, Script, 
          </div>
        </div>
        <div>
          <label>
            Publisher: 
          </label>
          <div
            class="book__container--header__info--publisher"
            data-testid="publisher"
          >
            Javascript
          </div>
        </div>
        <div>
          <label>
            Published Date: 
          </label>
          <div
            class="book__container--header__info--published"
            data-testid="published"
          >
            1980-10-10
          </div>
        </div>
      </div>
    </div>
    <div
      class="book__container--description"
    >
      Learn Javascript
    </div>
    <div
      class="book__container--button__container"
    >
      <button
        data-testid="addToWishList"
        type="button"
      >
        Add to wishlist
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BookSearch Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="book__search"
  >
    <form
      class="book__search--search__form"
    >
      <fieldset>
        <legend>
          Search Books
        </legend>
        <label
          for="book_search"
        >
          Book Name
        </label>
        <input
          aria-label="bookSearch"
          class="book__search--search__input"
          id="book_search"
          name="book search"
          placeholder="Search for books to add to your reading list and press Enter"
          type="search"
          value=""
        />
      </fieldset>
    </form>
    <div
      class="book__search--container"
    />
  </div>
</DocumentFragment>
`;
